### ISSUE TITLE
PNG and GIF images can't be cropped anymore

### ISSUE THREAD

**Issue by gmclelland** [processwire/processwire-issues#465]
### Short description of the issue

Since PW 3.0.86+ people are reporting they can't crop png images anymore.  See https://processwire.com/talk/topic/18141-pw-3086-png-images-cant-be-cropped-anymore/

### Expected behavior

When inserting a PNG image into a text area, I expected to be able to crop it.

### Actual behavior

The cropped image isn't saving after you make your crop selection and hit the apply button.

### Optional: Screenshots/Links that demonstrate the issue
Making a crop selection:
![cropping](https://user-images.githubusercontent.com/40570/34578735-b09b024a-f14b-11e7-91a6-c1096ffb5004.jpg)

After hitting the apply button:
![after-hit-apply-button](https://user-images.githubusercontent.com/40570/34578780-d7e3b6ee-f14b-11e7-9ac2-a4d401ba6483.jpg)

### Steps to reproduce the issue

1. Step 1
Drag an image into an image field.
2. Step 2
Insert the image into a text are via CKeditor image button
3. Step 3
Crop the image
4. Click Apply
Notice it doesn't save the cropped image

Note: Cropping works correctly for jpg images, just not png images.  No browser errors are listed in Chrome's inspector.

**Comment by gmclelland** [comment]
Updated issue title.  Can't resize GIF images as well.

**Comment by gmclelland** [comment]
I tested cropping on PW 3.0.85 with a PNG and GIF and it works.  So I guess the problem was introduced in 3.0.86.

**Comment by ryancramerdesign** [comment]
It looks like this recently added block of code is the culprit. If I comment it out, then PNG cropping starts working again: https://github.com/processwire/processwire/commit/bac60dc340d66766e7b9c0ce811a2ae152585123#diff-c649382351e78512ce2213adee9ad043R206

@horst-n Are you able to take a look at this? Is it okay to remove that block of code?

**Comment by horst-n** [comment]
@ryancramerdesign Let me have a look at it. We have added this as an issue / request where PNGs comes out with black backgounds. So, at this state only images that are to small for further processing, because upscaling is disallowed(!), gets outputted as the initial, unchanged image. So only thing that is the culprit here, is that the extracrop process was already applied, but no further processing is needed. If that's the case, we only have to add a flag to this and check it here. 
I will check it this weekend.

**Comment by ryancramerdesign** [comment]
Thanks Horst. I've commented it out temporarily for the moment, per this issue report. I will also take a closer look here shortly (I've been working with the ImageSizerEngines this week). But if you find a change I should apply just let me know. Thanks.

**Comment by gmclelland** [comment]
I can confirm PNG and GIFs are cropping now with 3.0.88.  Thanks for getting this back to a working state so quickly.

**Comment by horst-n** [comment]
@ryancramerdesign Hi Ryan, I have solved it.

A file with the fix is here: [ImageSizerEngineGD_(with_issue_fix_PNG_and_GIF).php](http://biriba.de/pw_pop3/ImageSizerEngineGD_(with_issue_fix_PNG_and_GIF).phps)

(search for "$isModified")


TL;DR:
In the function processResize() I first have added a local var $isModified:

`$this->modified = $isModified = false;  // added the local var $isModified to check for extra-crop manipulation`


Afterwards we need to check at three positions and optionally change its state:

`
		if($this->rotate || $needRotation) {
			$degrees = $this->rotate ? $this->rotate : $orientations[0];
			$image = $this->imRotate($image, $degrees);
			if(abs($degrees) == 90 || abs($degrees) == 270) {
				// we have to swap width & height now!
				$tmp = array($this->getWidth(), $this->getHeight());
				$this->setImageInfo($tmp[1], $tmp[0]);
				$isModified = true;                           // ****************** change new FLAG ******************
			}
		}

		if($this->flip || $needRotation) {
			$vertical = null;
			if($this->flip) {
				$vertical = $this->flip == 'v';
			} else if($orientations[1] > 0) {
				$vertical = $orientations[1] == 2;
			}
			if(!is_null($vertical)) {
				$image = $this->imFlip($image, $vertical);
				$isModified = true;                           // ****************** change new FLAG ******************
			}
		}

		// if there is requested to crop _before_ resize, we do it here @horst
		if(is_array($this->cropExtra)) {
			// check if we can load a second copy from sourceimage into ram
			if(self::checkMemoryForImage(array($this->info['width'], $this->info['height'], 3)) === false) {
				throw new WireException(basename($srcFilename) . " - not enough memory to load a copy for cropExtra");
			}

			$imageTemp = imagecreatetruecolor(imagesx($image), imagesy($image));  // create an intermediate memory image
			$this->prepareImageLayer($imageTemp, $image);
			imagecopy($imageTemp, $image, 0, 0, 0, 0, imagesx($image), imagesy($image)); // copy our initial image into the intermediate one
			imagedestroy($image); // release the initial image

			// get crop values and create a new initial image
			list($x, $y, $w, $h) = $this->cropExtra;

			// check if we can load a cropped version into ram
			if(self::checkMemoryForImage(array($w, $h, 3)) === false) {
				throw new WireException(basename($srcFilename) . " - not enough memory to load a cropped version for cropExtra");
			}

			$image = imagecreatetruecolor($w, $h);
			$this->prepareImageLayer($image, $imageTemp);
			imagecopy($image, $imageTemp, 0, 0, $x, $y, $w, $h);
			unset($x, $y, $w, $h);
			$isModified = true;                               // ****************** change new FLAG ******************

`


Finally we need to query and respect its state:

`		if(0 == $resizeMethod) {

			// this is the case if the original size is requested or a greater size but upscaling is set to false

			// current version is already the desired result, we only may have to compress JPEGs but leave GIF and PNG as is:
			if(!$isModified && ($this->imageType == \IMAGETYPE_PNG || $this->imageType == \IMAGETYPE_GIF)) {
`

**Comment by horst-n** [comment]
@ryancramerdesign ping. ?

**Comment by ryancramerdesign** [comment]
Thanks Horst. I actually did add this, but GitHub went down right as I was
replying to you last week, and I hadn't made it back yet. It'll be in this
weeks commits. Thanks for solving it.

**Comment by horst-n** [comment]
Great, thanks. A bit OT, but: will the module for animated gif handling I sent you a while back also make it into the core, now that the license issues are solved? (I PM'd you this via PW forum too) I need to know this, especially the date when it will go into the core or better, its final modules name a few days before it goes into the core. I have to update the legacy third party module to check for the new one in the core, so that it doesn't do a double processing of animated gifs in systems where both are installed after a wire-upgrade. 
https://github.com/jacoka/GIFDecoder/issues/1#issuecomment-353263850
https://github.com/xurei/GIFDecoder_optimized/issues/2#issuecomment-356342801

**Comment by ryancramerdesign** [comment]
@horst-n Sorry I got sidetracked, but I remember our conversation on this and glad to hear we can proceed with adding it. I'm doing some other work on the images stuff over the last couple of weeks, so will work on getting this added as part of the other stuff that's getting added. Thanks.

**Comment by horst-n** [comment]
@ryancramerdesign Yes I already thought this must be the reason. Also it tooks 2 month to get together the responses from the GIF EN-/DECODER authors.
If you don't have the older conversations at hand: the anim-gif engine must have higher priority than the default (fallback) GD engine. And it should be enabled by default.

The other image stuff currently going on is fine! Can't wait to see that "focal point" tool for responsive images across different dimensions integrated into the core! Unfortunately there is not much time for me atm to cowork. But if you like to get a second opinion or a tester for something new image related, feel free to drop me a note.

**Comment by ryancramerdesign** [comment]
@horst-n I added the animated GIF module yesterday. Is this one that should be automatically installed? Or do we leave it to the user to decide? I'm thinking auto-install, but wanted to double check.

**Comment by horst-n** [comment]
**@ryancramerdesign** auto-installed is perfect! Its default loading index number 9 is perfect too.
The only thing in regard of ImageRenderingEngines, (before the next stable version of PW goes public), is: I really want to see a check if Imagick is supported on a server within the installation routine, and if it is supported, it should be installed automatically too. This way we have the users best covered right from the start on.

**Comment by ryancramerdesign** [comment]
@horst-n Sounds good, I will set it to auto-install. Also regarding ImageSizerEngineIMagick: often times what's available (or not available) at install time might not be the same as when a site actually goes online. I'm wondering if the check for Imagick should be done at intervals instead, like post login maybe (where we already do some other system checks).

**Comment by szabeszg** [comment]
Since GD an Imagick produce different results I propose not to make any automatic switch from one to another – especially when when both are available – without first asking the superuser or at least notifying them. Otherwise supporting the possibility of an easy switch is welcome, of course.

**Comment by netcarver** [comment]
@ryancramerdesign @horst-n @szabeszg @gmclelland 

I see Ryan's tagged this as "fixed (close when ready)", do you think this issue has been adequately resolved? If there is a consensus, perhaps @gmclelland could close the issue. Many thanks!
